src/NuGet.Core/NuGet.ProjectModel/MemoryEfficientJsonPackageSpecReader.cs

```diff
@@ -0,0 +1,1570 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using Newtonsoft.Json;
+using NuGet.Common;
+using NuGet.Configuration;
+using NuGet.Frameworks;
+using NuGet.LibraryModel;
+using NuGet.Packaging.Core;
+using NuGet.RuntimeModel;
+using NuGet.Versioning;
+
+namespace NuGet.ProjectModel
+{
+    internal static class MemoryEfficientJsonPackageSpecReader
+    {
+        private static readonly char[] DelimitedStringSeparators = { ' ', ',' };
+        private static readonly char[] VersionSeparators = new[] { ';' };
+
+        internal static PackageSpec GetPackageSpec(JsonTextReader jsonReader, string packageSpecPath)
+        {
+            return GetPackageSpec(jsonReader, name: null, packageSpecPath, snapshotValue: null);
+        }
+
+        private static PackageSpec GetPackageSpec(JsonTextReader jsonReader, string name, string packageSpecPath, string snapshotValue)
+        {
+            var packageSpec = new PackageSpec();
+
+            string[] authors = null;
+            BuildOptions buildOptions = null;
+            List<CompatibilityProfile> compatibilityProfiles = null;
+            IList<string> contentFiles = null;
+            List<RuntimeDescription> runtimeDescriptions = null;
+            string title = null;
+            string version = null;
+            var wasPackOptionsSet = false;
+            var isMappingsNull = false;
```

Comment:
Can we have a positive condition name?

Reply:
It _is_ a positive condition name.  `isMappingsNotNull` would be a negative condition not a positive one.  A double negative (e.g.:  `if (!isMappingsNotNull) ...`) [has been avoided](https://github.com/NuGet/NuGet.Client/pull/3239/files#diff-639b72962d98b9eabacc47f0325923acR164).

I get the sense that you feel that "null" is an inherently negative name, but I don't think it is.  It can be part of a negative expression though:  `isNull` or `isNotNull`.  In fact, [`IsNull` is used commonly throughout the .NET Framework](https://referencesource.microsoft.com/#q=IsNull).

What I'm trying to do here is reproduce the same error condition that `JsonPackageSpecReader` produces when `packOptions`/`files`/`mappings` property exists but its value is not a valid object so `Mappings` should be set to `null`.

Reply:
Makes sense.